### ISSUE TITLE
Problem with creating classes from scratch with soot. 

### ISSUE THREAD

**Issue by eshza** [Sable/soot#918]
I am working on Mac OS and using soot to load all the classes from an apk and then write out the necessary classes to a jar file. 
This is how the class files are being created : 
public static void writeLibraryClassFile(SootClass cls) throws IOException {
		Options.v().set_output_dir(Paths.libraryClassesOutputDirectory().getPath());

		File file = new File(SourceLocator.v().getFileNameFor(cls, Options.output_format_class));
		file.getParentFile().mkdirs();

		OutputStream streamOut = new JasminOutputStream(new FileOutputStream(file));
		PrintWriter writerOut = new PrintWriter(new OutputStreamWriter(streamOut));

		if (cls.containsBafBody()) {
			 new soot.baf.JasminClass(cls).print(writerOut);
		} else {
			
			new soot.jimple.JasminClass(cls).print(writerOut);
		}

		writerOut.flush();
		streamOut.close();
	}
A problem occurs when creating class files with the same name, where one of the names is capitalized and one isn’t. 

For example, instead of creating two different files com.example.mchla.helloworld.view and com.example.mchla.helloworld.View, Soot just overwrites the file com.example.mchla.helloworld.view with the contents of the file com.example.mchla.helloworld.View.

**Comment by ericbodden** [comment]
Hello. This is not a problem of Soot but rather a misunderstanding on your side :-) The problem is that MacOS's file system is not really case sensitive. You cannot have two different files that only differ in the case. Hope that helps!

**Comment by karimhamdanali** [comment]
Thanks @ericbodden for getting back to us. However the classes `com.example.mchla.helloworld.view` and `com.example.mchla.helloworld.View` are not a real classes in the APK that @eshza is analyzing. They are both classes that Soot/FlowDroid model somehow internally to be able to precisely analyze the APK. This means that somehow, internally, Soot/FlowDroid is creating this class once with all lower-case name `view` and another one with the camel case `View`.

**Comment by mbenz89** [comment]
I guess this is some helper class for the callback analysis of Flowdroid. Hence, you might be better off asking on Flowdroid's issue tracker. Maybe @StevenArzt can help here.

However, I've never stumbled across this class when writing out classes that were loaded from an apk. It might get filtered out during the conventional write out process.

You can write out classes to a jar file like this:
```
Options.v().set_output_jar(true);
Options.v().set_output_dir("/some/dir");
Options.v().set_output_format(Options.output_format_class);

// depending on your needs, this might be useful
// Options.v().set_no_writeout_body_releasing(true);
// Options.v().set_force_overwrite(true);

PackManager.v().writeOutput();
```

**Comment by StevenArzt** [comment]
I don't know any reason why FlowDroid should create classes with such names. Have you checked that these classes are not phantoms? In case there are references to non-existing classes, e.g., from inside layout XML files, the normal Soot class resolver will create a phantom for exactly the class name that is being queried.

**Comment by karimhamdanali** [comment]
Thanks @mbenz89 for the suggestion. However, we won't be able to use that suggested method, because of the way we are doing our analyses.

@StevenArzt, both classes indeed result from parsing layout XML files (see stack trace below hone we set a breakpoint in the SootClass constructor). My question is how/why Soot queries these two classes  that only differ in the case and not the full class name?

```
		Thread [main] (Suspended (breakpoint at line 94 in SootClass))	
			SootClass.<init>(String, int) line: 94	
			SootClass.<init>(String) line: 119	
			SootResolver.makeClassRef(String) line: 116	
			SootResolver.resolveClass(String, int) line: 131	
			Scene.forceResolve(String, int) line: 1815	
			LayoutFileParser.getLayoutClass(String) line: 82	
			LayoutFileParser.parseLayoutNode(String, AXmlNode) line: 282	
			LayoutFileParser.parseLayoutNode(String, AXmlNode) line: 289	
			LayoutFileParser.parseLayoutNode(String, AXmlNode) line: 289	
			LayoutFileParser.parseLayoutNode(String, AXmlNode) line: 289	
			LayoutFileParser.access$4(LayoutFileParser, String, AXmlNode) line: 251	
			LayoutFileParser$2.handleResourceFile(String, Set<String>, InputStream) line: 235	
			LayoutFileParser(AbstractResourceParser).handleAndroidResourceFiles(String, Set<String>, IResourceHandler) line: 61	
			LayoutFileParser.parseLayoutFileDirect(String) line: 199	
			LayoutFileParser$1.internalTransform(String, Map) line: 183	
			LayoutFileParser$1(SceneTransformer).transform(String, Map<String,String>) line: 39	
			Transform.apply() line: 90	
			ScenePack.internalApply() line: 40	
			ScenePack(Pack).apply() line: 118	
			SetupApplication.calculateCallbackMethods(LayoutFileParser, SootClass) line: 683	
			SetupApplication.calculateCallbacks(ISourceSinkDefinitionProvider, SootClass) line: 490	
			SetupApplication.calculateCallbacks(ISourceSinkDefinitionProvider) line: 458	
			SetupApplication.runInfoflow(ISourceSinkDefinitionProvider) line: 1355	
			SetupApplication.constructCallgraph() line: 1182	
			SetupAndroid.getDummyMainMethod() line: 90	
			Main.main(String[]) line: 105
```

**Comment by StevenArzt** [comment]
We take the class names from the layout XML files as they are written there. If the layout XML files contain a reference to some class "view" and to some other class "View", that's how it is. We must assume that there are two different implementations of view classes that shall become part of the Activity's UI. Maybe that's a bug in the app - or some special case in the layout XML format that we have overlooked. It would be great if you could decompile the app using apktool and have a look at the respective layout files to see how they are structured. This might shed some light on whether we misinterpret some element as a class reference or whether the app is just weird.

**Comment by eshza** [comment]
@StevenArzt, I decompiled the apk as you suggested and found that one of the xml files contain a reference to ‘view’ whereas another one to ‘View’. However, I don’t think this is a bug in the apk. I have looked at some  other apks and it seems as though ‘view’ and ‘View’ are being used interchangeably. It is my understanding that they are both the same but for some reason the layout XML format allows you to use a lower case v.
